peer/persistentpeer_mgr.go

```diff
@@ -0,0 +1,479 @@
+)
+
+const (
+	// defaultMultiAddrConnectionStagger is the number of seconds to wait
+	// between attempting to a peer with each of its advertised addresses.
+	defaultMultiAddrConnectionStagger = 10 * time.Second
+)
+
+type PersistentPeerMgrConfig struct {
+	// ConnMgr is used to manage the creation and removal of connection
+	// requests. It handles the actual connection to a peer.
+	ConnMgr connMgr
+
+	// SubscribeTopology will be used to listen for updates to a persistent
+	// peer's advertised addresses.
+	SubscribeTopology func() (*routing.TopologyClient, error)
+
+	ChainNet wire.BitcoinNet
+
+	MultiAddrConnectionStagger time.Duration
+}
+
+// PersistentPeerManager manages persistent peers and the active connection
+// requests we may have to them.
+type PersistentPeerManager struct {
+	cfg *PersistentPeerMgrConfig
+
+	// conns maps a peer's public key string to a persistentPeer object.
+	conns map[string]*persistentPeer
+
+	// connsMu is used to guard the conns map.
+	connsMu sync.RWMutex
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+// persistentPeer holds the addresses we stored for a persistent peer along
+// with any active connection requests that we have for them. A persistentPeer
+// is permanent if we want to maintain a connection to them even when we have
+// no open channels with them.
+type persistentPeer struct {
+	// addrs is all the addresses we know about for this peer.
+	addrs map[string]*lnwire.NetAddress
+
+	// connReqs holds all the active connection requests that we have
+	// for the peer.
+	connReqs []*connmgr.ConnReq
+
+	// perm indicates if we should maintain a connection with a peer even
+	// if we have no channels with the peer.
+	perm bool
+
+	// retryCanceller is used to cancel any retry attempts with backoffs
+	// that are still maturing.
+	retryCanceller *chan struct{}
+}
+
+// connMgr is what the PersistentPeerManager will use to create and remove
+// connection requests. The purpose of this interface is to make testing easier.
+type connMgr interface {
+	Connect(c *connmgr.ConnReq)
+	Remove(id uint64)
+}
+
+// NewPersistentPeerManager creates a new PersistentPeerManager instance.
+func NewPersistentPeerManager(
+	cfg *PersistentPeerMgrConfig) *PersistentPeerManager {
+
+	if cfg.MultiAddrConnectionStagger == 0 {
+		cfg.MultiAddrConnectionStagger = defaultMultiAddrConnectionStagger
+	}
+
+	return &PersistentPeerManager{
+		cfg:   cfg,
+		conns: make(map[string]*persistentPeer),
+		quit:  make(chan struct{}),
+	}
+}
+
+// Start begins the processes of PersistentPeerManager. It subscribes to
+// graph updates and listens for any NodeAnnouncement messages that indicate
+// that the addresses of one of the persistent peers has changed and then
+// updates the peer's addresses and connection requests accordingly.
+func (m *PersistentPeerManager) Start() error {
+	graphSub, err := m.cfg.SubscribeTopology()
+	if err != nil {
+		return err
+	}
+
+	m.wg.Add(1)
+	go func() {
+		defer func() {
+			graphSub.Cancel()
```

Comment:
because the subscription must not be canceled when `Start` exits, it must be canceled when the goroutine exits. Start will exit as soon as it has spun off the goroutine